### PR TITLE
Backport to 9_4_X of  TensorFlow-based DeepDoubleB tagger

### PR DESCRIPTION
**Pull request by jmduarte**
Backport of #23033 so the DeepDoubleB tagger may be used in 9_4_X. The tagger is disabled in MINIAOD by default.

Rebased and locally tested over `CMSSW_9_4_X_2018-06-10-0000`

### CODE REVIEW

**Comment by slava77** [comment]
@cmsbuild please test with cms-sw/cmsdist#4113

File: PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+    # slimmedJets with DeepFlavour (remove DeepFlavour-less)
+    delattr(process, 'slimmedJets')
+    addToProcessAndTask('slimmedJets', getattr(process,'selectedUpdatedPatJetsSlimmedDeepFlavour'+postfix).clone(), process, task)
+    # delete module not used anymore (slimmedJets substitutes)
+    delattr(process, 'selectedUpdatedPatJetsSlimmedDeepFlavour'+postfix)
+
+
+
+    ## update slimmed jets to include DeepFlavour (keep same name)
```
Comment (slava77) [inline PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py]: I think that here it's more practical to fold this with an if
```python
 #switch to True if needed for private tests
enableDoubleB = False
if enableDoubleB:
...
```

File: RecoBTag/TensorFlow/BuildFile.xml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+<use name="FWCore/Utilities"/>
+<use name="PhysicsTools/TensorFlow"/>
+<use name="TrackingTools/IPTools"/>
+<use name="DataFormats/ParticleFlowCandidate"/>
+<use name="DataFormats/PatCandidates"/>
```
Comment (slava77) [inline RecoBTag/TensorFlow/BuildFile.xml]: These DataFormats were not needed in the master version of this file.
Was it really needed here?

**Comment by slava77** [comment]
ping
there are two unresolved comments in this PR review.

File: RecoBTag/TensorFlow/BuildFile.xml (hunk 2 above)
Comment (jmduarte) [reply RecoBTag/TensorFlow/BuildFile.xml]: Yes, I'm not entirely sure what changed, but when I remove the two `DataFormats` lines, I get compilation errors like:

```
>> Building shared library tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/libRecoBTagTensorFlow.so
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/TrackInfoBuilder.o: In function `btagbtvdeep::TrackInfoBuilder::buildTrackInfo(reco::Candidate const*, ROOT::Math::DisplacementVector3D<ROOT::Math::Cartesian3D<double>, ROOT::Math::DefaultCoordinateSystemTag> const&, Vector3DBase<float, GlobalTag>, reco::Vertex const&)':
TrackInfoBuilder.cc:(.text+0x14e): undefined reference to `typeinfo for pat::PackedCandidate'
TrackInfoBuilder.cc:(.text+0x162): undefined reference to `typeinfo for reco::PFCandidate'
TrackInfoBuilder.cc:(.text+0x1ca): undefined reference to `reco::PFCandidate::trackRef() const'
TrackInfoBuilder.cc:(.text+0x1eb): undefined reference to `reco::PFCandidate::trackRef() const'
TrackInfoBuilder.cc:(.text+0x20e): undefined reference to `reco::PFCandidate::trackRef() const'
TrackInfoBuilder.cc:(.text+0x8dc): undefined reference to `reco::PFCandidate::gsfTrackRef() const'
TrackInfoBuilder.cc:(.text+0x8fd): undefined reference to `reco::PFCandidate::gsfTrackRef() const'
TrackInfoBuilder.cc:(.text+0x920): undefined reference to `reco::PFCandidate::gsfTrackRef() const'
TrackInfoBuilder.cc:(.text+0x1792): undefined reference to `pat::PackedCandidate::unpackTrk() const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/TrackInfoBuilder.o: In function `pat::PackedCandidate::pseudoTrack() const':
TrackInfoBuilder.cc:(.text._ZNK3pat15PackedCandidate11pseudoTrackEv[_ZNK3pat15PackedCandidate11pseudoTrackEv]+0x19): undefined reference to `pat::PackedCandidate::unpackTrk() const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/TrackInfoBuilder.o: In function `reco::PFCandidate::bestTrack() const':
TrackInfoBuilder.cc:(.text._ZNK4reco11PFCandidate9bestTrackEv[_ZNK4reco11PFCandidate9bestTrackEv]+0x39): undefined reference to `reco::PFCandidate::trackRef() const'
TrackInfoBuilder.cc:(.text._ZNK4reco11PFCandidate9bestTrackEv[_ZNK4reco11PFCandidate9bestTrackEv]+0x54): undefined reference to `reco::PFCandidate::trackRef() const'
TrackInfoBuilder.cc:(.text._ZNK4reco11PFCandidate9bestTrackEv[_ZNK4reco11PFCandidate9bestTrackEv]+0x6f): undefined reference to `reco::PFCandidate::trackRef() const'
TrackInfoBuilder.cc:(.text._ZNK4reco11PFCandidate9bestTrackEv[_ZNK4reco11PFCandidate9bestTrackEv]+0x169): undefined reference to `reco::PFCandidate::gsfTrackRef() const'
TrackInfoBuilder.cc:(.text._ZNK4reco11PFCandidate9bestTrackEv[_ZNK4reco11PFCandidate9bestTrackEv]+0x184): undefined reference to `reco::PFCandidate::gsfTrackRef() const'
TrackInfoBuilder.cc:(.text._ZNK4reco11PFCandidate9bestTrackEv[_ZNK4reco11PFCandidate9bestTrackEv]+0x1a4): undefined reference to `reco::PFCandidate::gsfTrackRef() const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/ShallowTagInfoConverter.o: In function `btagbtvdeep::bTagToFeatures(reco::TaggingVariableList const&, btagbtvdeep::ShallowTagInfoFeatures&)':
ShallowTagInfoConverter.cc:(.text+0x1a): undefined reference to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const'
ShallowTagInfoConverter.cc:(.text+0x33): undefined reference to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const'
ShallowTagInfoConverter.cc:(.text+0x4d): undefined reference to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const'
ShallowTagInfoConverter.cc:(.text+0x67): undefined reference to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const'
ShallowTagInfoConverter.cc:(.text+0x81): undefined reference to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/ShallowTagInfoConverter.o:ShallowTagInfoConverter.cc:(.text+0x9b): more undefined references to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const' follow
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/ShallowTagInfoConverter.o: In function `btagbtvdeep::bTagToFeatures(reco::TaggingVariableList const&, btagbtvdeep::ShallowTagInfoFeatures&)':
ShallowTagInfoConverter.cc:(.text+0xe6): undefined reference to `reco::TaggingVariableList::getList(reco::btau::TaggingVariableName, bool) const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/BoostedDoubleSVTagInfoConverter.o: In function `btagbtvdeep::doubleBTagToFeatures(reco::TaggingVariableList const&, btagbtvdeep::BoostedDoubleSVTagInfoFeatures&)':
BoostedDoubleSVTagInfoConverter.cc:(.text+0x1a): undefined reference to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const'
BoostedDoubleSVTagInfoConverter.cc:(.text+0x33): undefined reference to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const'
BoostedDoubleSVTagInfoConverter.cc:(.text+0x4d): undefined reference to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const'
BoostedDoubleSVTagInfoConverter.cc:(.text+0x67): undefined reference to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const'
BoostedDoubleSVTagInfoConverter.cc:(.text+0x81): undefined reference to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/BoostedDoubleSVTagInfoConverter.o:BoostedDoubleSVTagInfoConverter.cc:(.text+0x9b): more undefined references to `reco::TaggingVariableList::get(reco::btau::TaggingVariableName, float) const' follow
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/NeutralCandidateConverter.o: In function `btagbtvdeep::packedCandidateToFeatures(pat::PackedCandidate const*, pat::Jet const&, float, float, btagbtvdeep::NeutralCandidateFeatures&)':
NeutralCandidateConverter.cc:(.text+0x1c8): undefined reference to `pat::PackedCandidate::puppiWeight() const'
NeutralCandidateConverter.cc:(.text+0x241): undefined reference to `pat::PackedCandidate::unpack() const'
NeutralCandidateConverter.cc:(.text+0x254): undefined reference to `pat::PackedCandidate::unpack() const'
NeutralCandidateConverter.cc:(.text+0x270): undefined reference to `pat::PackedCandidate::unpack() const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/NeutralCandidateConverter.o: In function `pat::PackedCandidate::pt() const':
NeutralCandidateConverter.cc:(.text._ZNK3pat15PackedCandidate2ptEv[_ZNK3pat15PackedCandidate2ptEv]+0x21): undefined reference to `pat::PackedCandidate::unpack() const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/NeutralCandidateConverter.o: In function `pat::PackedCandidate::phi() const':
NeutralCandidateConverter.cc:(.text._ZNK3pat15PackedCandidate3phiEv[_ZNK3pat15PackedCandidate3phiEv]+0x21): undefined reference to `pat::PackedCandidate::unpack() const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/NeutralCandidateConverter.o:NeutralCandidateConverter.cc:(.text._ZNK3pat15PackedCandidate3etaEv[_ZNK3pat15PackedCandidate3etaEv]+0x21): more undefined references to `pat::PackedCandidate::unpack() const' follow
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/ChargedCandidateConverter.o: In function `btagbtvdeep::packedCandidateToFeatures(pat::PackedCandidate const*, pat::Jet const&, btagbtvdeep::TrackInfoBuilder const&, float, float, btagbtvdeep::ChargedCandidateFeatures&)':
ChargedCandidateConverter.cc:(.text+0x42): undefined reference to `pat::PackedCandidate::puppiWeight() const'
ChargedCandidateConverter.cc:(.text+0x15c): undefined reference to `pat::PackedCandidate::unpackTrk() const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/ChargedCandidateConverter.o: In function `btagbtvdeep::recoCandidateToFeatures(reco::PFCandidate const*, reco::Jet const&, btagbtvdeep::TrackInfoBuilder const&, float, float, float, int, edm::Ref<std::vector<reco::Vertex, std::allocator<reco::Vertex> >, reco::Vertex, edm::refhelper::FindUsingAdvance<std::vector<reco::Vertex, std::allocator<reco::Vertex> >, reco::Vertex> > const&, btagbtvdeep::ChargedCandidateFeatures&)':
ChargedCandidateConverter.cc:(.text+0x1dd): undefined reference to `reco::PFCandidate::trackRef() const'
ChargedCandidateConverter.cc:(.text+0x1f8): undefined reference to `reco::PFCandidate::trackRef() const'
ChargedCandidateConverter.cc:(.text+0x38b): undefined reference to `reco::PFCandidate::trackRef() const'
ChargedCandidateConverter.cc:(.text+0x3a9): undefined reference to `reco::PFCandidate::trackRef() const'
ChargedCandidateConverter.cc:(.text+0x3c4): undefined reference to `reco::PFCandidate::trackRef() const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/ChargedCandidateConverter.o:ChargedCandidateConverter.cc:(.text+0x42c): more undefined references to `reco::PFCandidate::trackRef() const' follow
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/ChargedCandidateConverter.o: In function `btagbtvdeep::recoCandidateToFeatures(reco::PFCandidate const*, reco::Jet const&, btagbtvdeep::TrackInfoBuilder const&, float, float, float, int, edm::Ref<std::vector<reco::Vertex, std::allocator<reco::Vertex> >, reco::Vertex, edm::refhelper::FindUsingAdvance<std::vector<reco::Vertex, std::allocator<reco::Vertex> >, reco::Vertex> > const&, btagbtvdeep::ChargedCandidateFeatures&)':
ChargedCandidateConverter.cc:(.text+0x7a9): undefined reference to `reco::PFCandidate::gsfTrackRef() const'
ChargedCandidateConverter.cc:(.text+0x7c4): undefined reference to `reco::PFCandidate::gsfTrackRef() const'
ChargedCandidateConverter.cc:(.text+0x7df): undefined reference to `reco::PFCandidate::gsfTrackRef() const'
ChargedCandidateConverter.cc:(.text+0x899): undefined reference to `reco::PFCandidate::gsfTrackRef() const'
ChargedCandidateConverter.cc:(.text+0x8b7): undefined reference to `reco::PFCandidate::gsfTrackRef() const'
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/ChargedCandidateConverter.o:ChargedCandidateConverter.cc:(.text+0x8d2): more undefined references to `reco::PFCandidate::gsfTrackRef() const' follow
tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/ChargedCandidateConverter.o: In function `void btagbtvdeep::commonCandidateToFeatures<pat::PackedCandidate>(pat::PackedCandidate const*, reco::Jet const&, btagbtvdeep::TrackInfoBuilder const&, float const&, float const&, btagbtvdeep::ChargedCandidateFeatures&)':
ChargedCandidateConverter.cc:(.text._ZN11btagbtvdeep25commonCandidateToFeaturesIN3pat15PackedCandidateEEEvPKT_RKN4reco3JetERKNS_16TrackInfoBuilderERKfSE_RNS_24ChargedCandidateFeaturesE[_ZN11btagbtvdeep25commonCandidateToFeaturesIN3pat15PackedCandidateEEEvPKT_RKN4reco3JetERKNS_16TrackInfoBuilderERKfSE_RNS_24ChargedCandidateFeaturesE]+0x2a6): undefined reference to `pat::PackedCandidate::unpack() const'
collect2: error: ld returned 1 exit status
config/SCRAM/GMake/Makefile.rules:2068: recipe for target 'tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/libRecoBTagTensorFlow.so' failed
gmake: *** [tmp/slc6_amd64_gcc630/src/RecoBTag/TensorFlow/src/RecoBTagTensorFlow/libRecoBTagTensorFlow.so] Error 1
gmake: *** [There are compilation/build errors. Please see the detail log above.] Error 2
```

Comment (slava77) [reply RecoBTag/TensorFlow/BuildFile.xml]: ok
nothing to do for this file then.

File: PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py (hunk 1 above)
Comment (jmduarte) [reply PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py]: done in eb073bb2b978713a0244a5c2e62cd49ded51c346

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
backport of #23033

**Comment by slava77** [comment]
+1

for https://github.com/cms-sw/cmssw/pull/23582	eb073bb
- backport from the master version is correct, apart from disabling the double-b tagger in PhysicsTools/PatAlgos/python/slimming/applyDeepBtagging_cff.py in order to preserve compatibility with existing workflows in 94X. The double-b tagger can be enabled at analysis level if desired.
- jenkins tests pass and comparisons with the baseline show no differences as expected

**Comment by fabiocos** [comment]
+1

**Comment by fabiocos** [comment]
merge

**Pull request merged**